drivers/net/usb/smsc95xx.c

```diff
@@ -926,51 +929,86 @@ static int smsc95xx_ioctl(struct net_device *netdev, struct ifreq *rq, int cmd)
 	return generic_mii_ioctl(&dev->mii, if_mii(rq), cmd, NULL);
 }
 
+/* Increment the non-OUI part of the MAC address */
+static int smsc95xx_increment_lastmac(u8 *lastmac)
+{
+	spin_lock(&macaddr_lock);
+	if (last_macaddr[5] == 0xff) {
+		if (last_macaddr[4] == 0xff) {
+			if (last_macaddr[3] == 0xff) {
+				spin_unlock(&macaddr_lock);
+				return -1;
+			}
+			last_macaddr[3]++;
+		}
+		last_macaddr[4]++;
+	}
+	last_macaddr[5]++;
+
+	spin_unlock(&macaddr_lock);
+	return 0;
+}
+
 /* Check the macaddr module parameter for a MAC address */
 static int smsc95xx_macaddr_param(struct usbnet *dev, u8 *dev_mac)
 {
-	int i, j, got_num, num;
+	int i, j, got_num, num, ret;
 	u8 mtbl[MAC_ADDR_LEN];
 
 	if (macaddr[0] == ':')
 		return -1;
 
-	i = 0;
-	j = 0;
-	num = 0;
-	got_num = 0;
-	while (j < MAC_ADDR_LEN) {
-		if (macaddr[i] && macaddr[i] != ':') {
-			got_num++;
-			if ('0' <= macaddr[i] && macaddr[i] <= '9')
-				num = num * 16 + macaddr[i] - '0';
-			else if ('A' <= macaddr[i] && macaddr[i] <= 'F')
-				num = num * 16 + 10 + macaddr[i] - 'A';
-			else if ('a' <= macaddr[i] && macaddr[i] <= 'f')
-				num = num * 16 + 10 + macaddr[i] - 'a';
-			else
+	spin_lock(&macaddr_lock);
+	ret = is_valid_ether_addr(last_macaddr);
+	spin_unlock(&macaddr_lock);
+	if (!ret) {
+		i = 0;
```

Comment:
There are possible race conditions: Two threads could check is_valid_ether_addr() at the same time, both coming to the result that it is not valid yet and then both trying to change it in the following code. 
I suggest to lock the whole function smsc95xx_macaddr_param(). You could also use a mutex for this since AFAICS we are nowhere in atomic context and this not on a hot path (as a rule of thumb use sleepable locks like mutexes whenever possible and use spinlocks only if mutexes/semaphores cant be used).